mathics/builtin/numbers/hyperbolic.py

```diff
@@ -0,0 +1,348 @@
+# -*- coding: utf-8 -*-
+
+"""
+Hyperbolic Functions
+
+<url>:Hyperbolic functions: https://en.wikipedia.org/wiki/Hyperbolic_functions</url> are analogues of the ordinary trigonometric functions, but defined using the hyperbola rather than the circle.
+
+Numerical values and derivatives can be computed; however, most special exact values and simplification rules are not implemented yet.
+"""
+
+from typing import Optional
+from mathics.core.convert.sympy import SympyExpression
+
+from mathics.builtin.arithmetic import _MPMathFunction
+from mathics.builtin.base import Builtin
+from mathics.core.atoms import IntegerM1
+from mathics.core.expression import Expression
+from mathics.core.symbols import Symbol, SymbolPower
+
+SymbolArcCosh = Symbol("ArcCosh")
+SymbolArcSinh = Symbol("ArcSinh")
+SymbolCosh = Symbol("Cosh")
+SymbolSinh = Symbol("Sinh")
+
+
+class ArcCosh(_MPMathFunction):
+    """
+    <dl>
+      <dt>'ArcCosh[$z$]'
+      <dd>returns the inverse hyperbolic cosine of $z$.
+    </dl>
+
+    >> ArcCosh[0]
+     = I / 2 Pi
+    >> ArcCosh[0.]
+     = 0. + 1.5708 I
+    >> ArcCosh[0.00000000000000000000000000000000000000]
+     = 1.5707963267948966192313216916397514421 I
+    #> ArcCosh[1.4]
+     = 0.867015
+    """
+
+    summary_text = "inverse hyperbolic cosine function"
+    sympy_name = "acosh"
+    mpmath_name = "acosh"
+
+    rules = {
+        "Derivative[1][ArcCosh]": "1/(Sqrt[#-1]*Sqrt[#+1])&",
+    }
+
+
+class ArcCoth(_MPMathFunction):
+    """
+    <dl>
+      <dt>'ArcCoth[$z$]'
+      <dd>returns the inverse hyperbolic cotangent of $z$.
+    </dl>
+
+    >> ArcCoth[0]
+     = I / 2 Pi
+    >> ArcCoth[1]
+     = Infinity
+    >> ArcCoth[0.0]
+     = 0. + 1.5708 I
+    >> ArcCoth[0.5]
+     = 0.549306 - 1.5708 I
+
+    #> ArcCoth[0.000000000000000000000000000000000000000]
+     = 1.57079632679489661923132169163975144210 I
+    """
+
+    summary_text = "inverse hyperbolic cotangent function"
+    sympy_name = "acoth"
+    mpmath_name = "acoth"
+
+    rules = {
+        "ArcCoth[z:0.0]": "N[I / 2 Pi, Precision[1+z]]",
+        "Derivative[1][ArcCoth]": "1/(1-#^2)&",
+    }
+
+
+class ArcCsch(_MPMathFunction):
+    """
+    <dl>
+      <dt>'ArcCsch[$z$]'
+      <dd>returns the inverse hyperbolic cosecant of $z$.
+    </dl>
+
+    >> ArcCsch[0]
+     = ComplexInfinity
+    >> ArcCsch[1.0]
+     = 0.881374
+    """
+
+    mpmath_name = "acsch"
+
+    rules = {
+        "ArcCsch[0]": "ComplexInfinity",
+        "ArcCsch[0.]": "ComplexInfinity",
+        "Derivative[1][ArcCsch]": "-1 / (Sqrt[1+1/#^2] * #^2) &",
+    }
+
+    summary_text = "inverse hyperbolic cosecant function"
+    sympy_name = ""
+
+    def to_sympy(self, expr, **kwargs) -> Optional[SympyExpression]:
```

Comment:
Why this is necessary here (and in the next built-in) and not in the other?

Reply:
I don't understand. Please be more specific.

Reply:
I mean, why ArcCoth don't need to_sympy?

Reply:
ArcCoth is implemented as purely term rewriting. There is no evaluation method. Hey - if we can't fix the backwards function naming convention because it is pervasive, let's please agree _not_ to call them "apply" methods but "evaluation" or "eval" methods since that is what they are and that is what the rest of the informed world calls it.

Reply:
The above is not quite correct - sorry. `ArcCoth` uses the generic SymPy evaluation method and that presumably handles `to_sympy`, for ArcCsc someone decided to code in a transformation rule as python code right before passing it off to sympy.

Reply:
That makes sense - thanks.